WebCodecs decoder frames format is always I420

youennf opened the pull request:
#### d0d68fb2f7179d3c1ee7eedc1615455a8a619456
<pre>
WebCodecs decoder frames format is always I420
<a href="https://rdar.apple.com/145731406">rdar://145731406</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288711">https://bugs.webkit.org/show_bug.cgi?id=288711</a>

Reviewed by Jer Noble.

We were reporting all decoded frames to be I420, which is false.
For VP9 p2 in software, frames are 10 bits.
Given we do not yet support I420P10, we should expose null instead of &apos;I420&apos;.
Update code accordingly.

* LayoutTests/http/wpt/webcodecs/vp9-p2-decoder-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/vp9-p2-decoder.html: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::configure):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameFromOtherFrame):
(WebCore::WebCodecsVideoFrame::initializeFrameWithResourceAndSize):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
* Source/WebCore/platform/VideoPixelFormat.h:

Canonical link: <a href="https://commits.webkit.org/291499@main">https://commits.webkit.org/291499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee222eec2b3131ed455976aa93ce9e8e2e43eacb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12611 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/2351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98059 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12892 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21066 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71158 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/disable_controls_reposition.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28574 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96062 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9717 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84219 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51486 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9411 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1853 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42899 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79702 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1831 "Unable to confirm if test failures are introduced by change, retrying build") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100085 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14752 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80181 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79483 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19736 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24029 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1349 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13192 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20096 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25272 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19783 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23243 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21524 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -284,24 +284,25 @@ static std::optional<Exception> validateI420Sizes(const WebCodecsVideoFrame::Buf
 
     DOMRectInit defaultRect { 0, 0, static_cast<double>(init.codedWidth), static_cast<double>(init.codedHeight) };
-    auto parsedRectOrExtension = parseVisibleRect(defaultRect, init.visibleRect, init.codedWidth, init.codedHeight, init.format);
+    auto parsedRectOrExtension = parseVisibleRect(defaultRect, init.visibleRect, init.codedWidth, init.codedHeight, *init.format);
```
Can we replace this unconditional dereference with a `.value_or()`?

youennf replied:
This method is called from JS binding, where format should never be nullopt, hence the ASSERT and the unconditional dereference.
I'll add a local var.

jernoble commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -284,24 +284,25 @@ static std::optional<Exception> validateI420Sizes(const WebCodecsVideoFrame::Buf
 
     auto parsedRect = parsedRectOrExtension.releaseReturnValue();
-    auto layoutOrException = computeLayoutAndAllocationSize(defaultRect, init.layout, init.format);
+    auto layoutOrException = computeLayoutAndAllocationSize(defaultRect, init.layout, *init.format);
```
Ditto.

jernoble commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -284,24 +284,25 @@ static std::optional<Exception> validateI420Sizes(const WebCodecsVideoFrame::Buf
         return Exception { ExceptionCode::TypeError, makeString("Data is too small "_s, data.length(), " / "_s, layout.allocationSize) };
 
-    auto colorSpace = videoFramePickColorSpace(init.colorSpace, init.format);
+    auto colorSpace = videoFramePickColorSpace(init.colorSpace, *init.format);
```
Ditto.

jernoble reviewed (approved):
r=me with nits

youennf committed:
WebCodecs decoder frames format is always I420
rdar://145731406
https://bugs.webkit.org/show_bug.cgi?id=288711

Reviewed by Jer Noble.

We were reporting all decoded frames to be I420, which is false.
For VP9 p2 in software, frames are 10 bits.
Given we do not yet support I420P10, we should expose null instead of 'I420'.
Update code accordingly.

* LayoutTests/http/wpt/webcodecs/vp9-p2-decoder-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/vp9-p2-decoder.html: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::configure):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameFromOtherFrame):
(WebCore::WebCodecsVideoFrame::initializeFrameWithResourceAndSize):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
* Source/WebCore/platform/VideoPixelFormat.h:

Canonical link: https://commits.webkit.org/291499@main
